pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -7952,6 +8045,351 @@ spec:
+      - safari
+      - firefox
+  serviceAccountName: test-sa
+  pipelineRef:
+    name: p-dag
+status:
+  pipelineSpec:
+    params:
+     - name: platforms
+       type: array
+     - name: browsers
+       type: array
+    tasks:
+      - name: platforms-and-browsers
+        matrix:
+          params:
+            - name: platform
+              value:
+               - linux
+               -  mac
+               -  windows
+            - name: browser
+              value:
+               - chrome
+               - safari
+               - firefox
+        taskRef:
+          name: mytask
+          kind: Task
+  conditions:
+  - type: Succeeded
+    status: "Unknown"
+    reason: "Running"
+    message: "Tasks Completed: 0 (Failed: 0, Cancelled 0), Incomplete: 1, Skipped: 0"
+  childReferences:
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-0
+    pipelineTaskName: platforms-and-browsers
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-1
+    pipelineTaskName: platforms-and-browsers
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-2
+    pipelineTaskName: platforms-and-browsers
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-3
+    pipelineTaskName: platforms-and-browsers
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-4
+    pipelineTaskName: platforms-and-browsers
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-5
+    pipelineTaskName: platforms-and-browsers
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-6
+    pipelineTaskName: platforms-and-browsers
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-7
+    pipelineTaskName: platforms-and-browsers
+  - apiVersion: tekton.dev/v1beta1
+    kind: TaskRun
+    name: pr-platforms-and-browsers-8
+    pipelineTaskName: platforms-and-browsers
+`),
+	}}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			pr := parse.MustParseV1beta1PipelineRun(t, fmt.Sprintf(`
+metadata:
+  name: pr
+  namespace: foo
+spec:
+  serviceAccountName: test-sa
+  params:
+  - name: platforms
+    value:
+     - linux
+     - mac
+     - windows
+  - name: browsers
+    value:
+     - chrome
+     - safari
+     - firefox
+  pipelineRef:
+    name: %s
+`, tt.name))
+			d := test.Data{
+				PipelineRuns: []*v1beta1.PipelineRun{pr},
+				Pipelines:    []*v1beta1.Pipeline{tt.p},
+				Tasks:        []*v1beta1.Task{task},
+				ConfigMaps:   cms,
+			}
+			if tt.tr != nil {
+				d.TaskRuns = []*v1beta1.TaskRun{tt.tr}
+			}
+			prt := newPipelineRunTest(t, d)
+			defer prt.Cancel()
+
+			_, clients := prt.reconcileRun("foo", "pr", []string{}, false)
+			taskRuns, err := clients.Pipeline.TektonV1beta1().TaskRuns("foo").List(prt.TestAssets.Ctx, metav1.ListOptions{
```

Comment:
Maybe? I copied the same format as out existing Matrix Tests

https://github.com/tektoncd/pipeline/blob/7d6cd810e2ea4348415be9b094d1db9deb625e3e/pkg/reconciler/pipelinerun/pipelinerun_test.go#L7924

Reply:
I think this helper function would be better for a few reasons: it's a bit more modular, it doesn't have to be updated if the names of the tasks are changed, it's much clearer what it does, and it doesn't introduce the mysterious magic number Limit = 1. I suspect the reason this is present in the existing task is to exclude the unmatrixed pipeline taskrun from the results, but it's not clear. As an example of how this could lead to unintended results, what if you added a new pipeline task to the test case, and there was a bug that caused the taskruns to not get generated? This test would still pass since it's filtering for a specific pipeline task.

Reply:
Update to use helper function